### PR TITLE
Leak fixes

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

**Comment by Therzok** [comment]
The commits should now be atomic and easy to review separately.

File: main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.Wrappers/SemanticHighlightingSyntaxMode.cs [hunk 1]
```diff
@@ -163,7 +163,15 @@ public void Dispose()
 				var segments = new List<ColoredSegment> (syntaxLine.Segments);
 				int endOffset = segments [segments.Count - 1].EndOffset;
 				try {
-					var tree = lineSegments.FirstOrDefault (t => t.Item1 == line);
+					Tuple<IDocumentLine, HighlightingSegmentTree> tree = null;
+
+					// This code should not have any lambda capture linq, as it is a hot loop.
+					foreach (var segment in lineSegments) {
+						if (segment.Item1 == line) {
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.Wrappers/SemanticHighlightingSyntaxMode.cs]: @mkrueger I'm still not sure how you want me to modify this.

Comment (Therzok) [inline main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.Wrappers/SemanticHighlightingSyntaxMode.cs]: Do we want to do a linenumber check or?

Comment (mkrueger) [inline main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.Wrappers/SemanticHighlightingSyntaxMode.cs]: Basically that's the only option here however.
segment.Item1.Offset == line.Offset will do it. It's slower than the reference comparison but reference comparsion of lines doesn't work anymore.

**Comment by alanmcgovern** [comment]
This PR breaks code highlighting nearly completely for me. I stopped testing pretty much immediately once I noticed that. It might be easier to split these up into smaller PRs which can be tested and merged separately.

**Comment by alanmcgovern** [comment]
Screencast: https://screencast.com/t/yQLrK47L

**Comment by Therzok** [comment]
Highlighting is broken in master, not because of this Pr

File: main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.Wrappers/SemanticHighlightingSyntaxMode.cs (hunk 1 above)
Comment (Therzok) [inline main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.Wrappers/SemanticHighlightingSyntaxMode.cs]: If a whole line is deleted, the offsets equality check would return the wrong line, would it not?

**Comment by Therzok** [comment]
Rebased on top of origin/master to get the highlighting fix in.

File: main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.Wrappers/SemanticHighlightingSyntaxMode.cs (hunk 1 above)
Comment (mkrueger) [inline main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.Wrappers/SemanticHighlightingSyntaxMode.cs]: Yes but then it should be corrected in the next semantic highlighting update.

But with the current model segment.Item1 == line is always false.

Comment (Therzok) [inline main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.Wrappers/SemanticHighlightingSyntaxMode.cs]: IDocumentLine implements ISegments which implements IEquatable. Can't we just change to `Equals`?

**Comment by alanmcgovern** [comment]
This is 👍 for me. I ran with the commits all day and it seems fine. The highlighting issue was a separate regression that's been fixed already :)

**Comment by Therzok** [comment]
Lane created, waiting for results.

**Comment by Therzok** [comment]
We have:
* Orange external addins - not this PR's fault
* Orange indentation tests - this PR hasn't been rebased on top of indentation test fixes from Mike

Other than that, it's green.

**Pull request merged**